samza-api/src/main/java/org/apache/samza/metrics/ListGauge.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.metrics;
+
+import java.util.Collection;
+import java.util.LinkedList;
+
+
+/**
+ * A ListGauge is a collection of {@link org.apache.samza.metrics.Gauge}. ListGauges are useful for maintaining, recording,
+ * or collecting specific Gauge values over time. It is implemented as a {@link org.apache.samza.metrics.Metric}.
+ * For example, the set of recent errors that have occurred.
+ *
+ * This current implementation uses a size-bound-policy and holds the N most-recent Gauge objects added to the list.
+ * This bound N is configurable at instantiation time.
+ * TODO: Support a time-based and size-and-time-based hybrid policy.
```

Comment:
If this PR is only adding gauges, maybe move the TODOs to another JIRA and assign them to yourself. You can remove these from code.

Reply:
done